text/3627-match-ergonomics-2024.md

```diff
@@ -0,0 +1,337 @@
+- Feature Name: `ref_pat_eat_one_layer_2024`
+- Start Date: 2024-05-06
+- RFC PR: [rust-lang/rfcs#3627](https://github.com/rust-lang/rfcs/pull/3627)
+- Rust Issue: [rust-lang/rust#123076](https://github.com/rust-lang/rust/issues/123076)
+
+# Summary
+[summary]: #summary
+
+Various changes to the match ergonomics rules:
+
+- On edition ≥ 2024, `&` and `&mut` patterns only remove a single layer of
+  references.
+- On edition ≥ 2024, `mut` on an identifier pattern does not force its binding
+  mode to by-value.
+- On edition ≥ 2024, `&` patterns can match against `&mut` references.
+- On all editions, the binding mode can no longer ever be implicitly set to
+  `ref mut` behind an `&` pattern.
+
+# Motivation
+[motivation]: #motivation
+
+Match ergonomics have been a great success overall, but there are some surprising
+interactions that regularly confuse users.
+
+- `mut` resets the binding mode to by-value, which users do not expect; the
+  mutability of the binding seems like a separate concern from its type
+  (<https://github.com/rust-lang/rust/issues/105647>,
+  <https://github.com/rust-lang/rust/issues/112545>)
+- `&` and `&mut` patterns must correspond with a reference in the same position
+  in the scrutinee, even if there is an inherited reference present. Therefore,
+  users have no general mechanism to "cancel out" an inherited reference
+  (<https://users.rust-lang.org/t/reference-of-tuple-and-tuple-of-reference/91713/6>,
+  <https://users.rust-lang.org/t/cannot-deconstruct-reference-inside-match-on-reference-why/92147>,
+  <https://github.com/rust-lang/rust/issues/50008>,
+  <https://github.com/rust-lang/rust/issues/64586>)
+- When an `&` or `&mut` pattern is used in a location where there is also an
+  inherited reference present, both are stripped; adding a single `&` to the
+  pattern can remove two `&`s from the type of the binding.
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+Match ergonomics works a little differently in edition 2024 and above.
+
+## Matching against inherited references
+
+In all editions, when you match against an `&` or `&mut` reference with the type
+of its referent, you get an "inherited reference": the binding mode of
+"downstream" bindings is set to `ref` or `ref mut`.
+
+```rust
+// Unchanged from old editions:
+// `x` "inherits" the `&` from the scrutinee type.
+let [x] = &[42];
+let _: &u8 = x;
+```
+
+In edition 2024 and above, an `&` or `&mut` pattern can match against this
+inherited reference, consuming it. A pattern that does this has no other effect.
+
+```rust
+// New in edition 2024:
+// `&` pattern consumes inherited `&` reference.
+let [&x] = &[42];
+let _: u8 = x;
+```
+
+## `&` matches against `&mut`
+
+In edition 2024 and above, `&` patterns can match against `&mut` references
+(including "inherited" references).
+
+```rust
+let &foo = &mut 42;
```

Comment:
`foo` is still immutable right?

```rust
let &foo = &mut 42;
foo = 53; // error[E0384]: cannot assign twice to immutable variable `foo`
```

Reply:
Yes. (It would be immutable with `let &mut foo = &mut 42;` also.)